central/policycategory/datastore/datastore_impl.go

```diff
@@ -32,11 +35,81 @@ var (
 )
 
 type datastoreImpl struct {
-	storage  store.Store
-	indexer  index.Indexer
-	searcher search.Searcher
+	storage              store.Store
+	indexer              index.Indexer
+	searcher             search.Searcher
+	policyCategoryEdgeDS datastore.DataStore
+	categoryMutex        sync.Mutex
+}
+
+func (ds *datastoreImpl) SetPolicyCategoriesForPolicy(ctx context.Context, policyID string, categoryNames []string) error {
+	ds.categoryMutex.Lock()
+	defer ds.categoryMutex.Unlock()
+
+	results, err := ds.policyCategoryEdgeDS.SearchEdges(ctx, searchPkg.NewQueryBuilder().AddExactMatches(searchPkg.PolicyID, policyID).ProtoQuery())
+	if err != nil {
+		return err
+	}
+	existingCategoryNames := make([]string, 0, len(results))
+	for _, result := range results {
+		existingCategoryNames = append(existingCategoryNames, result.Name)
+	}
+	existingCategories := set.NewFrozenStringSet(existingCategoryNames...)
+	categoriesToUpdate := set.NewFrozenStringSet(categoryNames...)
+	if len(existingCategories.Difference(categoriesToUpdate).AsSlice()) == 0 &&
+		len(categoriesToUpdate.Difference(existingCategories).AsSlice()) == 0 {
+		// no edges to update since existing categories and categories to be updated are the same
+		return nil
+	}
+
+	// disassociate all categories from given policy
+	if err = ds.policyCategoryEdgeDS.DeleteByQuery(ctx, searchPkg.NewQueryBuilder().AddExactMatches(searchPkg.PolicyID, policyID).ProtoQuery()); err != nil {
+		return err
+	}
+	if len(categoryNames) == 0 {
+		return nil
+	}
+
+	// set associations to provided categories
+	categories := make([]*storage.PolicyCategory, 0, len(categoryNames))
+	err = ds.storage.Walk(policyCategoryCtx, func(category *storage.PolicyCategory) error {
+		categories = append(categories, category)
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	categoryIds := make([]string, 0, len(categoryNames))
+	categoryNameIDMap := categoryUtils.GetCategoryNameToIDs(categories)
+	categoriesToAdd := make([]*storage.PolicyCategory, 0)
+	for _, c := range categoryNames {
+		if categoryNameIDMap[strings.Title(c)] != "" {
+			categoryIds = append(categoryIds, categoryNameIDMap[c])
+		} else {
+			newCategory := &storage.PolicyCategory{
+				Id:        uuid.NewV4().String(),
+				Name:      strings.Title(c),
+				IsDefault: false,
+			}
+			categoriesToAdd = append(categoriesToAdd, newCategory)
```

Comment:
In incoming `categoryNames` is not deduped. You must dedupe it: `categoryNameIDMap[newCategory.Name] = newCategory.Id`